theano/sandbox/mkl/mkl_type.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import absolute_import, print_function, division
+import os
+import numpy
+import theano
+from theano import Type, Variable
+from theano import tensor
+from six import StringIO
+
+try:
+    import mkl_ndarray
+    import mkl_ndarray.mkl_ndarray as mkl
+
+    try:
+        mkl_ndarray.__file__
+    except AttributeError:
+        from theano.gof.cmodule import get_lib_extension
+        mkl_ndarray.__file__ = os.path.join(mkl_ndarray.__path__._path[0], 'mkl_ndarray.' + get_lib_extension())
+except ImportError:
+    mkl = None
+
+
+class _operators(tensor.basic._tensor_py_operators):
+    dtype = property(lambda s: s.type.dtype)
+    broadcastable = property(lambda s: s.type.broadcastable)
+    ndim = property(lambda s: s.type.ndim)
+
+
+class MKLNdarrayVariable(_operators, Variable):
+    pass
+
+
+class MKLNdarrayType(Type):
+    context_num = 'mkl'
+    Variable = MKLNdarrayVariable
+    Constant = None
+    SharedVariable = None
+    ndim = None
+    broadcastable = None
+
+    def __init__(self, broadcastable, name=None, dtype=None):
+        if dtype not in ('float32', 'float64'):
+            raise TypeError('%s only supports dtype float32/float64 for now. \
+                            Tried sing dtype float32 for variable %s' % (self.__class__.__name__, name))
+        self.dtype = dtype
+        self.typenum = 11 if dtype is 'float32' else 12
+        self.broadcastable = tuple(bool(b) for b in broadcastable)
+        self.name = name
+        self.dtype_specs()  # error checking is done there
+
+    def clone(self, dtype=None, broadcastable=None):
+        if broadcastable is None:
+            broadcastable = self.broadcastable
+        return self.__class__(broadcastable, name=self.name, dtype=dtype)
+
+    def filter(self, data, strict=False, allow_downcast=None):
+        # return self.filter_inplace(data, None, strict=strict, allow_downcast=allow_downcast)
+        return data
```

Comment:
The goal of filter is to cast data into a data represented by this type. Currently it don't do that.